{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import requests\n",
    "import xml.etree.ElementTree as ET\n",
    "import subprocess\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################\n",
    "# NOMS ET TXID DES BACTERIES D'INTERET #\n",
    "########################################\n",
    "# importation du fichier contenant les txid des bactéries d'intérêt (fichier obtenu via https://www.ncbi.nlm.nih.gov/Taxonomy/TaxIdentifier/tax_identifier.cgi)\n",
    "recup_txid = pd.read_csv('../sources/bacteria_of_interest/bacteria_of_interest_with_txids.txt', sep = '\\t')\n",
    "recup_txid['taxid']\n",
    "\n",
    "# replacement des taxid non identifiés par np.nan\n",
    "recup_txid['taxid'] = recup_txid['taxid'].replace({\" \" : np.nan})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 597 entries, 0 to 705\n",
      "Data columns (total 7 columns):\n",
      " #   Column          Non-Null Count  Dtype \n",
      "---  ------          --------------  ----- \n",
      " 0   code            597 non-null    int64 \n",
      " 1   |               597 non-null    object\n",
      " 2   name            597 non-null    object\n",
      " 3   |.1             597 non-null    object\n",
      " 4   preferred name  597 non-null    object\n",
      " 5   |.2             597 non-null    object\n",
      " 6   taxid           597 non-null    object\n",
      "dtypes: int64(1), object(6)\n",
      "memory usage: 37.3+ KB\n"
     ]
    }
   ],
   "source": [
    "# suppression des taxid non identifiés et des doublons\n",
    "recup_txid = recup_txid.dropna()\n",
    "recup_txid.drop_duplicates(subset = ['taxid'], keep = 'first', inplace = True)\n",
    "recup_txid.info()\n",
    "\n",
    "df_bacteria_of_interest = pd.DataFrame(data = {'name' : recup_txid['name'],\n",
    "                                               'taxid' : recup_txid['taxid']})\n",
    "\n",
    "df_bacteria_of_interest.reset_index(drop = True, inplace = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#exportation des txids dans un fichier qui servira à importer des séquences 16S correspondantes depuis NCBI\n",
    "txids = df_bacteria_of_interest['taxid'].to_list()\n",
    "\n",
    "with open(\"../sources/bacteria_of_interest/txids.txt\", \"w\") as txid_file:\n",
    "    for txid in txids:\n",
    "        txid_file.write(txid + '\\n')\n",
    "txid_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/marthe/miniconda3/envs/microbiote/lib/python3.11/site-packages/Bio/SeqIO/FastaIO.py:203: BiopythonDeprecationWarning: Previously, the FASTA parser silently ignored comments at the beginning of the FASTA file (before the first sequence).\n",
      "\n",
      "Nowadays, the FASTA file format is usually understood not to have any such comments, and most software packages do not allow them. Therefore, the use of comments at the beginning of a FASTA file is now deprecated in Biopython.\n",
      "\n",
      "In a future Biopython release, this deprecation warning will be replaced by a ValueError. To avoid this, there are three options:\n",
      "\n",
      "(1) Modify your FASTA file to remove such comments at the beginning of the file.\n",
      "\n",
      "(2) Use SeqIO.parse with the 'fasta-pearson' format instead of 'fasta'. This format is consistent with the FASTA format defined by William Pearson's FASTA aligner software. Thie format allows for comments before the first sequence; lines starting with the ';' character anywhere in the file are also regarded as comment lines and are ignored.\n",
      "\n",
      "(3) Use the 'fasta-blast' format. This format regards any lines starting with '!', '#', or ';' as comment lines. The 'fasta-blast' format may be safer than the 'fasta-pearson' format, as it explicitly indicates which lines are comments. \n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "#####################################################\n",
    "# SEQUENCES REPRÉSENTATIVES DES BACTÉRIES D'INTERÊT #\n",
    "#####################################################\n",
    "# Importation de séquences pour les bactéries d'intérêt depuis NCBI, à partir de leur txid\n",
    "\n",
    "Entrez.email = \"marthe.leoz@gmail.com\"\n",
    "\n",
    "for txid in txids[:]:\n",
    "\n",
    "    # Recherche des séquences d'ARN 16S\n",
    "    handle = Entrez.esearch(db=\"nucleotide\", term=f\"txid{txid}[All Fields] AND 16S[All Fields] AND rRNA[All Fields] AND partial[All Fields] NOT whole[All Fields] AND (\\\"1300\\\"[SLEN] : \\\"1500\\\"[SLEN]) \", retmax=10)\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "\n",
    "    # Obtenir les identifiants des séquences\n",
    "    id_list = record[\"IdList\"]\n",
    "\n",
    "    # Récupérer les séquences et les enregistrer dans un fichier de type fasta\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=id_list, rettype=\"fasta\", retmode=\"text\")\n",
    "    sequences = SeqIO.parse(handle, \"fasta\")\n",
    "\n",
    "    # Sauvegarder dans un fichier\n",
    "    filepath = \"../sources/bacteria_of_interest/bacteria_of_interest_sequences_max10.txt\" \n",
    "    with open(filepath, \"a\") as output_file:\n",
    "        # on inscrit chaque txid avant la série de séquences correspondantes pour pouvoir les reporter plus tard dans le dataframe\n",
    "        # le fichier est donc sauvegardé au format txt et non fasta\n",
    "        output_file.write('TXID : ' + str(txid) + '\\n')     \n",
    "        SeqIO.write(sequences, output_file, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>txid</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>dna_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>202789</td>\n",
       "      <td>&gt;MH645801.1 Actinobaculum massiliense strain F...</td>\n",
       "      <td>CCTTTTTTTGGTTGGGTNCTCGAGTGGCGAACGGTGAGTATTACGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>202789</td>\n",
       "      <td>&gt;MH279688.1 Actinobaculum massiliense strain F...</td>\n",
       "      <td>TTTTTGGTTGGGTGCTNGAGTGGCGAACGGGTGAGTATTACGTGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>202789</td>\n",
       "      <td>&gt;LT558806.1 Actinobaculum massiliense partial ...</td>\n",
       "      <td>AGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>202789</td>\n",
       "      <td>&gt;FJ711191.1 Actinobaculum massiliense strain N...</td>\n",
       "      <td>ACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGCCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>33007</td>\n",
       "      <td>&gt;PQ788148.1 Winkia neuii strain som 201 16S ri...</td>\n",
       "      <td>CGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGGATCCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     txid                                             seq_id  \\\n",
       "0  202789  >MH645801.1 Actinobaculum massiliense strain F...   \n",
       "1  202789  >MH279688.1 Actinobaculum massiliense strain F...   \n",
       "2  202789  >LT558806.1 Actinobaculum massiliense partial ...   \n",
       "3  202789  >FJ711191.1 Actinobaculum massiliense strain N...   \n",
       "4   33007  >PQ788148.1 Winkia neuii strain som 201 16S ri...   \n",
       "\n",
       "                                             dna_seq  \n",
       "0  CCTTTTTTTGGTTGGGTNCTCGAGTGGCGAACGGTGAGTATTACGT...  \n",
       "1  TTTTTGGTTGGGTGCTNGAGTGGCGAACGGGTGAGTATTACGTGAG...  \n",
       "2  AGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACA...  \n",
       "3  ACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGCCCT...  \n",
       "4  CGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGGATCCA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# création d'un df avec identifiants NCBI, txid, et séquences\n",
    "# lecture du fichier contenant les séquences regroupées par txids\n",
    "fichier = open('../sources/bacteria_of_interest/bacteria_of_interest_sequences_max10.txt', 'r')\n",
    "data = fichier.readlines()\n",
    "\n",
    "#initialisation du dataframe\n",
    "df_bacteria_of_interest_with_seq = pd.DataFrame(columns = ['txid', 'seq_id', 'dna_seq'])\n",
    "\n",
    "#initialisation d'une donnée (identifiant et séquence)\n",
    "txid = ''\n",
    "dna_seq = ''\n",
    "seq_id = ''\n",
    "\n",
    "#récupération d'un identifiant (caractérisé par le caractère '>' en début de ligne)\n",
    "#et de la séquence correspondante en concaténant les chaînes séparées par des retours à la ligne\n",
    "n = 0\n",
    "for ligne in data[:]:\n",
    "    if ligne[:7] == 'TXID : ':\n",
    "        new_txid = ligne[7:-1]\n",
    "    else:\n",
    "        if ligne[0] == '>' :\n",
    "            df_bacteria_of_interest_with_seq.loc[n] = [txid, seq_id, dna_seq]\n",
    "            txid = new_txid\n",
    "            seq_id = ligne[:-1]\n",
    "            dna_seq = ''\n",
    "            n += 1\n",
    "        else:\n",
    "            dna_seq = dna_seq + ligne[:-1]\n",
    "    \n",
    "\n",
    "#suppression de la première ligne (vide) du dataframe \n",
    "df_bacteria_of_interest_with_seq = df_bacteria_of_interest_with_seq[1:].reset_index(drop = True)\n",
    "# suppression des séquences identiques\n",
    "df_bacteria_of_interest_with_seq = df_bacteria_of_interest_with_seq.drop_duplicates(subset = 'dna_seq')\n",
    "\n",
    "#affichage de contrôle\n",
    "display(df_bacteria_of_interest_with_seq.head())\n",
    "\n",
    "#exportation en .csv\n",
    "df_bacteria_of_interest_with_seq.to_csv('../sources/bacteria_of_interest/bacteria_of_interest_sequences_max10.csv', sep = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#####################################\n",
    "# TAXONOMIE DES BACTÉRIES D'INTERÊT #\n",
    "#####################################\n",
    "# exportation des txids dupliqués pour récupérer les taxonomies via NCBI\n",
    "txids = df_bacteria_of_interest_with_seq['txid'].to_list()\n",
    "\n",
    "with open(\"../sources/bacteria_of_interest/txids_max10.txt\", \"w\") as txid_file:\n",
    "    for txid in txids:\n",
    "        txid_file.write(txid + '\\n')\n",
    "txid_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 4290 entries, 0 to 4289\n",
      "Data columns (total 3 columns):\n",
      " #   Column   Non-Null Count  Dtype \n",
      "---  ------   --------------  ----- \n",
      " 0   txid     4290 non-null   object\n",
      " 1   seq_id   4290 non-null   object\n",
      " 2   dna_seq  4290 non-null   object\n",
      "dtypes: object(3)\n",
      "memory usage: 100.7+ KB\n"
     ]
    }
   ],
   "source": [
    "df_bacteria_of_interest_with_seq = df_bacteria_of_interest_with_seq.reset_index(drop = True)\n",
    "df_bacteria_of_interest_with_seq.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy_levels = ['domain', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "\n",
    "# Fonction de récupération des données de chaque espèce via NCBI\n",
    "def get_data_from_ncbi(txid):\n",
    "    Entrez.email = \"marthe.leoz@gmail.com\"\n",
    "\n",
    "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&id={txid}&retmode=xml\"\n",
    "    response = requests.get(url)\n",
    "        \n",
    "    if response.status_code == 200:\n",
    "        return response.text\n",
    "    else:\n",
    "        print(f\"Erreur pour le txid {txid}: {response.status_code}\")\n",
    "    \n",
    "# Fonction d'extraction des différents niveaux de taxonomie dans les données récupérées\n",
    "def get_taxo_dict_from_data(data):\n",
    "    try:\n",
    "        tree = ET.ElementTree(ET.fromstring(data))\n",
    "        root = tree.getroot()\n",
    "\n",
    "        # instanciation du dictionnaire\n",
    "        taxo_dict = {}\n",
    "        for rank in taxonomy_levels:\n",
    "            taxo_dict[rank] = 'undefined'    \n",
    "        # Extraction de la taxonomie\n",
    "        for taxon in root.iter('Taxon'):\n",
    "            name = taxon.find('ScientificName').text\n",
    "            rank = taxon.find('Rank').text\n",
    "            taxo_dict[rank] = name\n",
    "        return taxo_dict\n",
    "            \n",
    "    except Exception as e:\n",
    "        print(f\"Erreur lors de l'analyse du XML: {e}\")\n",
    "\n",
    "# Fonction d'ajout des niveaux de taxonomie dans le df\n",
    "def insert_taxo_in_df(df, txid, taxo_dict):\n",
    "    # index des lignes concernées\n",
    "    list_ids = df.loc[df['txid'] == txid].index\n",
    "    # ajout dans le df\n",
    "    for k in list_ids:\n",
    "        for rank in taxonomy_levels:\n",
    "            df.loc[k, rank] = taxo_dict[rank]\n",
    "\n",
    "# Récupération et ajout des niveaux de taxonomie pour chaque séquences du df, à partir de leur txid\n",
    "\n",
    "######### WARNING #####\n",
    "##### LA RECUPERATION DES TAXO A TENDANCE À PLANTER EN COURS DE ROUTE... \n",
    "##### AJOUTER UN TRY OU QQE CHOSE POUR QU'IL PUISSE REPRENDRE OÙ IL EN ÉTAIT?\n",
    "##### ON PEUT RÉCUPÉRER L'INDEX À PARTIR DUQUEL REPRENDRE AVEC list(txids).index('txid_renvoyé_comme_erreur'),\n",
    "##### IL FAUT ENSUITE LE PLACER COMME BORNE DE DÉPART DANS 'for k in range (...)'\n",
    "#######################\n",
    "txids = df_bacteria_of_interest_with_seq['txid'].unique()\n",
    "for k in range(len(txids)):\n",
    "    txid = txids[k]\n",
    "    data = get_data_from_ncbi(txid)\n",
    "    taxo_dict = get_taxo_dict_from_data(data)\n",
    "    insert_taxo_in_df(df_bacteria_of_interest_with_seq, txid, taxo_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 4290 entries, 0 to 4289\n",
      "Data columns (total 10 columns):\n",
      " #   Column   Non-Null Count  Dtype \n",
      "---  ------   --------------  ----- \n",
      " 0   txid     4290 non-null   object\n",
      " 1   seq_id   4290 non-null   object\n",
      " 2   dna_seq  4290 non-null   object\n",
      " 3   domain   4290 non-null   object\n",
      " 4   phylum   4290 non-null   object\n",
      " 5   class    4290 non-null   object\n",
      " 6   order    4290 non-null   object\n",
      " 7   family   4290 non-null   object\n",
      " 8   genus    4290 non-null   object\n",
      " 9   species  4290 non-null   object\n",
      "dtypes: object(10)\n",
      "memory usage: 335.3+ KB\n"
     ]
    }
   ],
   "source": [
    "# Vérification de la complétude du df\n",
    "df_bacteria_of_interest_with_seq.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################\n",
    "# EXTRACTION DES REGIONS V3V4 DE CHAQUE SÉQUENCE #\n",
    "##################################################\n",
    "# le fichier .txt contenant les séquences contient également des txids. On supprime ces txids pour obtenir\n",
    "# un fichier au format fasta.\n",
    "\n",
    "with open('../sources/bacteria_of_interest/bacteria_of_interest_max10.fasta', 'a') as fasta_file:\n",
    "    for seq_id, dna_seq in zip(df_bacteria_of_interest_with_seq['seq_id'], df_bacteria_of_interest_with_seq['dna_seq']):\n",
    "        fasta_file.write(seq_id + '\\n' + dna_seq + '\\n')\n",
    "fasta_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extraction des régions V3V4 des séquences à l'aide de l'outil extract regions 16S\n",
    "# les résultats sont inscrits dans un fichier .fasta\n",
    "\n",
    "script = '/home/marthe/Documents/DS/projet/extract_regions_16s/extract_regions'\n",
    "input_file = '../sources/bacteria_of_interest/bacteria_of_interest_max10.fasta'\n",
    "output_file = '../sources/bacteria_of_interest/bacteria_of_interest_V3V4_max10.fasta'\n",
    "\n",
    "subprocess.call([script, '-i', input_file, '-f', '341F', '-r', '806R', '-o', output_file, '-t', '4'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# L'extracteur de régions du gène 16S remplace les bases T par des bases U, ce qui rend les fichiers\n",
    "# incompatibles avec qiime. On remet donc à nouveau des bases T à la place des bases U\n",
    "\n",
    "def replace_U_with_T(file_path):\n",
    "    with open(file_path, \"r+\") as file:\n",
    "        #read the file contents\n",
    "        file_contents = file.read()\n",
    "        text_pattern = re.compile(\"U\")\n",
    "        file_contents = text_pattern.sub(\"T\", file_contents)\n",
    "        file.seek(0)\n",
    "        file.truncate()\n",
    "        file.write(file_contents)\n",
    "\n",
    "replace_U_with_T('../sources/bacteria_of_interest/bacteria_of_interest_V3V4_max10.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 4290 entries, 0 to 4289\n",
      "Data columns (total 11 columns):\n",
      " #   Column   Non-Null Count  Dtype \n",
      "---  ------   --------------  ----- \n",
      " 0   txid     4290 non-null   object\n",
      " 1   seq_id   4290 non-null   object\n",
      " 2   dna_seq  4290 non-null   object\n",
      " 3   domain   4290 non-null   object\n",
      " 4   phylum   4290 non-null   object\n",
      " 5   class    4290 non-null   object\n",
      " 6   order    4290 non-null   object\n",
      " 7   family   4290 non-null   object\n",
      " 8   genus    4290 non-null   object\n",
      " 9   species  4290 non-null   object\n",
      " 10  V3V4     4256 non-null   object\n",
      "dtypes: object(11)\n",
      "memory usage: 368.8+ KB\n"
     ]
    }
   ],
   "source": [
    "# intégration des restrictions aux régions V3V4 de chaque séquence dans le dataframe\n",
    "with open('../sources/bacteria_of_interest/bacteria_of_interest_V3V4_max10.fasta', 'r') as V3V4_file:\n",
    "    lines = V3V4_file.readlines()\n",
    "    for k in range(df_bacteria_of_interest_with_seq.shape[0]):\n",
    "        extract_seq = lines[2*k+1]\n",
    "        if len(extract_seq) > 250:          # on ne conserve que les séquences extraites d'une longueur suffisante\n",
    "            df_bacteria_of_interest_with_seq.loc[k, 'V3V4'] = lines[2*k+1][:-1]\n",
    "        else:\n",
    "            df_bacteria_of_interest_with_seq.loc[k, 'V3V4'] = np.nan\n",
    "\n",
    "df_bacteria_of_interest_with_seq.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Suppression des lignes vides (séquences pour lesquelles l'extraction n'a retourné aucun résultat)\n",
    "\n",
    "def sup_empty_lines(file_path):\n",
    "    results = \"\"\n",
    "    with open(file_path, 'r+') as file:\n",
    "        lines = file.readlines()\n",
    "        i = 0\n",
    "        while i < len(lines):\n",
    "            if (lines[i] != '\\n') & (lines[i+1] != '\\n'):   #on teste la ligne de l'identifiant et la ligne de la séquence\n",
    "                results = results + lines[i] + lines[i+1]\n",
    "            i += 2\n",
    "    file.close()\n",
    "    with open(file_path, 'w') as file:\n",
    "        file.write(results)\n",
    "    file.close()\n",
    "\n",
    "\n",
    "\n",
    "sup_empty_lines('../sources/bacteria_of_interest/bacteria_of_interest_V3V4_max10.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 4256 entries, 0 to 4255\n",
      "Data columns (total 11 columns):\n",
      " #   Column   Non-Null Count  Dtype \n",
      "---  ------   --------------  ----- \n",
      " 0   txid     4256 non-null   object\n",
      " 1   seq_id   4256 non-null   object\n",
      " 2   dna_seq  4256 non-null   object\n",
      " 3   domain   4256 non-null   object\n",
      " 4   phylum   4256 non-null   object\n",
      " 5   class    4256 non-null   object\n",
      " 6   order    4256 non-null   object\n",
      " 7   family   4256 non-null   object\n",
      " 8   genus    4256 non-null   object\n",
      " 9   species  4256 non-null   object\n",
      " 10  V3V4     4256 non-null   object\n",
      "dtypes: object(11)\n",
      "memory usage: 365.9+ KB\n"
     ]
    }
   ],
   "source": [
    "# suppression des lignes correspondantes dans le df\n",
    "df_bacteria_of_interest_with_seq.dropna(subset = ['V3V4'], axis = 0, inplace = True)\n",
    "df_bacteria_of_interest_with_seq.reset_index(drop = True, inplace = True)\n",
    "df_bacteria_of_interest_with_seq.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# suppression des virgules dans seq_id\n",
    "pattern = re.compile(r'>[^,]*')\n",
    "df_bacteria_of_interest_with_seq['seq_id'] = df_bacteria_of_interest_with_seq['seq_id'].apply(lambda x : pattern.search(x).group())\n",
    "\n",
    "# suppression des doublons\n",
    "df_bacteria_of_interest_with_seq.drop_duplicates(['dna_seq'], inplace = True)\n",
    "df_bacteria_of_interest_with_seq.reset_index(drop = True, inplace = True)\n",
    "df_bacteria_of_interest_with_seq.head()\n",
    "\n",
    "df_bacteria_of_interest_with_seq.to_csv(\"../datasets/train_sets/urinary_max10.csv\", index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 4256 entries, 0 to 4255\n",
      "Data columns (total 11 columns):\n",
      " #   Column   Non-Null Count  Dtype \n",
      "---  ------   --------------  ----- \n",
      " 0   txid     4256 non-null   int64 \n",
      " 1   seq_id   4256 non-null   object\n",
      " 2   dna_seq  4256 non-null   object\n",
      " 3   domain   4256 non-null   object\n",
      " 4   phylum   4256 non-null   object\n",
      " 5   class    4256 non-null   object\n",
      " 6   order    4256 non-null   object\n",
      " 7   family   4256 non-null   object\n",
      " 8   genus    4256 non-null   object\n",
      " 9   species  4256 non-null   object\n",
      " 10  V3V4     4256 non-null   object\n",
      "dtypes: int64(1), object(10)\n",
      "memory usage: 365.9+ KB\n"
     ]
    }
   ],
   "source": [
    "df_bacteria_of_interest_with_seq = pd.read_csv('../datasets/train_sets/urinary_max10.csv')\n",
    "\n",
    "df_bacteria_of_interest_with_seq.drop_duplicates(['dna_seq'], inplace = True)\n",
    "df_bacteria_of_interest_with_seq.reset_index(drop = True, inplace = True)\n",
    "df_bacteria_of_interest_with_seq.info()\n",
    "\n",
    "# Exportation du dataframe en csv\n",
    "df_bacteria_of_interest_with_seq.to_csv(\"../datasets/train_sets/urinary_max10.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bacteria_of_interest_with_seq = pd.read_csv('../datasets/train_sets/urinary_max10.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_817/2788675725.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  urinary_taxo_for_qiime.loc[i, 'Feature ID'] = feature_id[0]\n",
      "/tmp/ipykernel_817/2788675725.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  urinary_taxo_for_qiime.loc[i, 'Taxon'] = taxon\n",
      "/tmp/ipykernel_817/2788675725.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  urinary_taxo_for_qiime.drop('seq_id', axis = 1, inplace = True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature ID</th>\n",
       "      <th>Taxon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MH645801.1</td>\n",
       "      <td>D_0__Bacteria;D_1__Actinomycetota;D_2__Actinom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MH279688.1</td>\n",
       "      <td>D_0__Bacteria;D_1__Actinomycetota;D_2__Actinom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LT558806.1</td>\n",
       "      <td>D_0__Bacteria;D_1__Actinomycetota;D_2__Actinom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FJ711191.1</td>\n",
       "      <td>D_0__Bacteria;D_1__Actinomycetota;D_2__Actinom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PQ788148.1</td>\n",
       "      <td>D_0__Bacteria;D_1__Actinomycetota;D_2__Actinom...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Feature ID                                              Taxon\n",
       "0  MH645801.1  D_0__Bacteria;D_1__Actinomycetota;D_2__Actinom...\n",
       "1  MH279688.1  D_0__Bacteria;D_1__Actinomycetota;D_2__Actinom...\n",
       "2  LT558806.1  D_0__Bacteria;D_1__Actinomycetota;D_2__Actinom...\n",
       "3  FJ711191.1  D_0__Bacteria;D_1__Actinomycetota;D_2__Actinom...\n",
       "4  PQ788148.1  D_0__Bacteria;D_1__Actinomycetota;D_2__Actinom..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#################################################\n",
    "# création de fichiers de format adapté à qiime #\n",
    "#################################################\n",
    "# Fichier contenant les taxonomies\n",
    "urinary_taxo_for_qiime = df_bacteria_of_interest_with_seq[['seq_id']]\n",
    "\n",
    "\n",
    "feature_id_pattern = re.compile(r'>([\\S]*)')      # Pour récupérer les identifiants au format '>xxxxxxxxx.x'\n",
    "for i in range(urinary_taxo_for_qiime.shape[0]):\n",
    "    feature_id = feature_id_pattern.findall(urinary_taxo_for_qiime.loc[i, 'seq_id'])\n",
    "    urinary_taxo_for_qiime.loc[i, 'Feature ID'] = feature_id[0]\n",
    "\n",
    "    taxons = [df_bacteria_of_interest_with_seq.loc[i, rank] for rank in taxonomy_levels]\n",
    "    taxon = 'D_0__' + taxons[0] + ';D_1__' + taxons[1] + ';D_2__' + taxons[2] + ';D_3__' + taxons[3] + ';D_4__' + taxons[4] + ';D_5__' + taxons[5] + ';D_6__' + taxons[6] \n",
    "    urinary_taxo_for_qiime.loc[i, 'Taxon'] = taxon\n",
    "\n",
    "\n",
    "urinary_taxo_for_qiime.drop('seq_id', axis = 1, inplace = True)\n",
    "\n",
    "display(urinary_taxo_for_qiime.head())\n",
    "urinary_taxo_for_qiime.to_csv('../sources/bacteria_of_interest/qiime_urinary_taxo_max10.tsv', sep = '\\t', index = False)\n",
    "\n",
    "\n",
    "# Fichier contenant les séquences complètes\n",
    "with open('../sources/bacteria_of_interest/qiime_urinary_seq_max10.fasta', 'a') as fasta_file:\n",
    "    for i in range(urinary_taxo_for_qiime.shape[0]):\n",
    "        fasta_file.write('>' + urinary_taxo_for_qiime.loc[i, 'Feature ID'] + '\\n' + df_bacteria_of_interest_with_seq.loc[i, 'dna_seq'] + '\\n')\n",
    "fasta_file.close()\n",
    "\n",
    "# Fichier contenant les séquences restreintes aux régions V3V4\n",
    "with open('../sources/bacteria_of_interest/qiime_urinary_seq_V3V4_max10.fasta', 'a') as fasta_file:\n",
    "    for i in range(urinary_taxo_for_qiime.shape[0]):\n",
    "        fasta_file.write('>' + urinary_taxo_for_qiime.loc[i, 'Feature ID'] + '\\n' + df_bacteria_of_interest_with_seq.loc[i, 'V3V4'] + '\\n')\n",
    "fasta_file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>txid</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>dna_seq</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>V3V4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>202789</td>\n",
       "      <td>&gt;MH279688.1 Actinobaculum massiliense strain F...</td>\n",
       "      <td>TTTTTGGTTGGGTGCTNGAGTGGCGAACGGGTGAGTATTACGTGAG...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinomycetota</td>\n",
       "      <td>Actinomycetes</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Actinomycetaceae</td>\n",
       "      <td>Actinobaculum</td>\n",
       "      <td>Actinobaculum massiliense</td>\n",
       "      <td>CNTACGGGAGGCAGCAGTGGGGGATATTGCACAATGGGCGCAAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>202789</td>\n",
       "      <td>&gt;LT558806.1 Actinobaculum massiliense partial ...</td>\n",
       "      <td>AGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACA...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinomycetota</td>\n",
       "      <td>Actinomycetes</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Actinomycetaceae</td>\n",
       "      <td>Actinobaculum</td>\n",
       "      <td>Actinobaculum massiliense</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGGGGGATATTGCACAATGGGCGCAAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>202789</td>\n",
       "      <td>&gt;FJ711191.1 Actinobaculum massiliense strain N...</td>\n",
       "      <td>ACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGCCCT...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinomycetota</td>\n",
       "      <td>Actinomycetes</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Actinomycetaceae</td>\n",
       "      <td>Actinobaculum</td>\n",
       "      <td>Actinobaculum massiliense</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGGGGGATATTGCACAATGGGCGCAAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>33007</td>\n",
       "      <td>&gt;OR999579.1 Winkia neuii strain CNSY1 16S ribo...</td>\n",
       "      <td>GGCCTGCGGCGTGCTTACCATGCAAGTCGAACGGGATCCATTAGCG...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinomycetota</td>\n",
       "      <td>Actinomycetes</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Actinomycetaceae</td>\n",
       "      <td>Winkia</td>\n",
       "      <td>Winkia neuii</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGGGGGATATTGCACAATGGACGAAAGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>33007</td>\n",
       "      <td>&gt;OR260435.1 Winkia neuii strain 19 16S ribosom...</td>\n",
       "      <td>AACGGGTGAGTAACACGTGAGTAACCTGCCCTTTTCTTTGGGATAA...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinomycetota</td>\n",
       "      <td>Actinomycetes</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Actinomycetaceae</td>\n",
       "      <td>Winkia</td>\n",
       "      <td>Winkia neuii</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGGGGGATATTGCACAATGGACGNAAGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4251</th>\n",
       "      <td>1886637</td>\n",
       "      <td>&gt;PV739705.1 Delftia sp. strain APQ6-11 16S rib...</td>\n",
       "      <td>CGGTGAACAGGATCTTCGGACGCTGACGAGTGGCGAACGGGTGAGT...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>Comamonadaceae</td>\n",
       "      <td>Delftia</td>\n",
       "      <td>Delftia sp.</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGGGGAATTTTGGACAATGGGCGAAAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4252</th>\n",
       "      <td>1886637</td>\n",
       "      <td>&gt;PV579005.1 Delftia sp. strain KRMS06 16S ribo...</td>\n",
       "      <td>GGTAAGAGTTTGATCCTGGCTCAGATTGAACGCTGGCGGCATGCCT...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>Comamonadaceae</td>\n",
       "      <td>Delftia</td>\n",
       "      <td>Delftia sp.</td>\n",
       "      <td>CCTACGGGAGGCGCAGTGGGGAATTTTGGACAATGGGCGAAAGCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4253</th>\n",
       "      <td>1886637</td>\n",
       "      <td>&gt;PV450672.1 Delftia sp. strain KDMSBB08 16S ri...</td>\n",
       "      <td>GAAGCACCAGTACTTACGCGACGCTGGACTGAGGGGCGAACGGGTG...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>Comamonadaceae</td>\n",
       "      <td>Delftia</td>\n",
       "      <td>Delftia sp.</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGGGGAATTTTGGACAATTGGGCGAAAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4254</th>\n",
       "      <td>1886637</td>\n",
       "      <td>&gt;PV382267.1 Delftia sp. strain YC2177 16S ribo...</td>\n",
       "      <td>CGTTGCCGGCATGCTTACACATGCAGTCGAACGGTAACAGGTCTTC...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>Comamonadaceae</td>\n",
       "      <td>Delftia</td>\n",
       "      <td>Delftia sp.</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGGGGAATTTTGGACAATGGGCGAAAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4255</th>\n",
       "      <td>1886637</td>\n",
       "      <td>&gt;PV262677.1 Delftia sp. strain Y32 16S ribosom...</td>\n",
       "      <td>CCTTACCTGCAGTCGAACGGTAACAGGTCTTCGGACGCTGACGAGT...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>Comamonadaceae</td>\n",
       "      <td>Delftia</td>\n",
       "      <td>Delftia sp.</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGGGGAATTTTGGACAATGGGCGAAAGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3686 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         txid                                             seq_id  \\\n",
       "1      202789  >MH279688.1 Actinobaculum massiliense strain F...   \n",
       "2      202789  >LT558806.1 Actinobaculum massiliense partial ...   \n",
       "3      202789  >FJ711191.1 Actinobaculum massiliense strain N...   \n",
       "5       33007  >OR999579.1 Winkia neuii strain CNSY1 16S ribo...   \n",
       "6       33007  >OR260435.1 Winkia neuii strain 19 16S ribosom...   \n",
       "...       ...                                                ...   \n",
       "4251  1886637  >PV739705.1 Delftia sp. strain APQ6-11 16S rib...   \n",
       "4252  1886637  >PV579005.1 Delftia sp. strain KRMS06 16S ribo...   \n",
       "4253  1886637  >PV450672.1 Delftia sp. strain KDMSBB08 16S ri...   \n",
       "4254  1886637  >PV382267.1 Delftia sp. strain YC2177 16S ribo...   \n",
       "4255  1886637  >PV262677.1 Delftia sp. strain Y32 16S ribosom...   \n",
       "\n",
       "                                                dna_seq    domain  \\\n",
       "1     TTTTTGGTTGGGTGCTNGAGTGGCGAACGGGTGAGTATTACGTGAG...  Bacteria   \n",
       "2     AGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACA...  Bacteria   \n",
       "3     ACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGCCCT...  Bacteria   \n",
       "5     GGCCTGCGGCGTGCTTACCATGCAAGTCGAACGGGATCCATTAGCG...  Bacteria   \n",
       "6     AACGGGTGAGTAACACGTGAGTAACCTGCCCTTTTCTTTGGGATAA...  Bacteria   \n",
       "...                                                 ...       ...   \n",
       "4251  CGGTGAACAGGATCTTCGGACGCTGACGAGTGGCGAACGGGTGAGT...  Bacteria   \n",
       "4252  GGTAAGAGTTTGATCCTGGCTCAGATTGAACGCTGGCGGCATGCCT...  Bacteria   \n",
       "4253  GAAGCACCAGTACTTACGCGACGCTGGACTGAGGGGCGAACGGGTG...  Bacteria   \n",
       "4254  CGTTGCCGGCATGCTTACACATGCAGTCGAACGGTAACAGGTCTTC...  Bacteria   \n",
       "4255  CCTTACCTGCAGTCGAACGGTAACAGGTCTTCGGACGCTGACGAGT...  Bacteria   \n",
       "\n",
       "              phylum               class            order            family  \\\n",
       "1     Actinomycetota       Actinomycetes  Actinomycetales  Actinomycetaceae   \n",
       "2     Actinomycetota       Actinomycetes  Actinomycetales  Actinomycetaceae   \n",
       "3     Actinomycetota       Actinomycetes  Actinomycetales  Actinomycetaceae   \n",
       "5     Actinomycetota       Actinomycetes  Actinomycetales  Actinomycetaceae   \n",
       "6     Actinomycetota       Actinomycetes  Actinomycetales  Actinomycetaceae   \n",
       "...              ...                 ...              ...               ...   \n",
       "4251  Pseudomonadota  Betaproteobacteria  Burkholderiales    Comamonadaceae   \n",
       "4252  Pseudomonadota  Betaproteobacteria  Burkholderiales    Comamonadaceae   \n",
       "4253  Pseudomonadota  Betaproteobacteria  Burkholderiales    Comamonadaceae   \n",
       "4254  Pseudomonadota  Betaproteobacteria  Burkholderiales    Comamonadaceae   \n",
       "4255  Pseudomonadota  Betaproteobacteria  Burkholderiales    Comamonadaceae   \n",
       "\n",
       "              genus                    species  \\\n",
       "1     Actinobaculum  Actinobaculum massiliense   \n",
       "2     Actinobaculum  Actinobaculum massiliense   \n",
       "3     Actinobaculum  Actinobaculum massiliense   \n",
       "5            Winkia               Winkia neuii   \n",
       "6            Winkia               Winkia neuii   \n",
       "...             ...                        ...   \n",
       "4251        Delftia                Delftia sp.   \n",
       "4252        Delftia                Delftia sp.   \n",
       "4253        Delftia                Delftia sp.   \n",
       "4254        Delftia                Delftia sp.   \n",
       "4255        Delftia                Delftia sp.   \n",
       "\n",
       "                                                   V3V4  \n",
       "1     CNTACGGGAGGCAGCAGTGGGGGATATTGCACAATGGGCGCAAGCC...  \n",
       "2     CCTACGGGAGGCAGCAGTGGGGGATATTGCACAATGGGCGCAAGCC...  \n",
       "3     CCTACGGGAGGCAGCAGTGGGGGATATTGCACAATGGGCGCAAGCC...  \n",
       "5     CCTACGGGAGGCAGCAGTGGGGGATATTGCACAATGGACGAAAGTC...  \n",
       "6     CCTACGGGAGGCAGCAGTGGGGGATATTGCACAATGGACGNAAGTC...  \n",
       "...                                                 ...  \n",
       "4251  CCTACGGGAGGCAGCAGTGGGGAATTTTGGACAATGGGCGAAAGCC...  \n",
       "4252  CCTACGGGAGGCGCAGTGGGGAATTTTGGACAATGGGCGAAAGCCT...  \n",
       "4253  CCTACGGGAGGCAGCAGTGGGGAATTTTGGACAATTGGGCGAAAGC...  \n",
       "4254  CCTACGGGAGGCAGCAGTGGGGAATTTTGGACAATGGGCGAAAGCC...  \n",
       "4255  CCTACGGGAGGCAGCAGTGGGGAATTTTGGACAATGGGCGAAAGCC...  \n",
       "\n",
       "[3686 rows x 11 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bacteria_of_interest_with_seq.loc[df_bacteria_of_interest_with_seq['txid'].duplicated() == True]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nombre de txid différents :  527\n"
     ]
    }
   ],
   "source": [
    "# suppression des classes sous représentées\n",
    "test = df_bacteria_of_interest_with_seq.groupby('txid').count().reset_index()\n",
    "minors = test.loc[test['seq_id'] < 2]['txid'].values\n",
    "#print(minors)\n",
    "\n",
    "filtered_df = df_bacteria_of_interest_with_seq.copy()\n",
    "filtered_df = filtered_df.loc[filtered_df['txid'].apply(lambda x : x not in minors)]\n",
    "print('nombre de txid différents : ', len(filtered_df['txid'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_df.to_csv('../datasets/train_sets/urinary_max10_min2.csv', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "microbiote",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
